Plug-in new text parsing engine

janherich opened the pull request:
[comment]: # (Please replace ... with your information. Remove < and >)
[comment]: # (To auto-close issue on merge, please insert the related issue number after # i.e fixes #566)

Plugs-in new functions from `chat.models.message-content` for detecting formatting/actions information in message text content into mobile flow/rendering.
By doing that, a lot of logic is moved from view to event processing/subscriptions and as a free bonus, we get working public chat links on mobile.
The PR has also potential to speed things up quite significantly for older, slower devices, as the new text-parsing is on average much quicker then the old implementation + runs less frequently (by not being in the view). 

This is also pre-requisite for implementing `@mentions` functionality.

### Steps to test:
- General light regression testing of 1-1 and public chats with normal/formatted messages (containing links and tags)
- It would be worthwhile to check performance (receiving a lot of messages) on slower android devices as this PR has a potential to help it quite significantly, especially for older, slower devices
- Test desktop as well, as this PR unifies the logic for mobile/desktop text formatting/actions, the only caveat is that italic formatting doesn't work on desktop as that font style is currently not supported there, will be fixed later

status: ready

janherich committed:
Plug-in new text parsing engine

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [x] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status.html))?

rasom commented on src/status_im/ui/screens/chat/message/message.cljs:
```diff
@@ -178,27 +63,48 @@
                   :number-of-lines 5}
       text]]))
 
+(def ^:private kind->props-fn {:bold   (constantly {:style {:font-weight :bold}})
```
not a big deal, but any chance we can move the map to the next line? it would probably fit into 80 chars then :)

janherich replied:
Of course, will do :)

rasom commented on src/status_im/ui/screens/chat/message/message.cljs:
```diff
@@ -178,27 +63,48 @@
+                 [react/text
+                  (merge {:key idx}
+                         (apply merge-with merge
+                                (keep (comp #(when % (% text message)) kind->props-fn) kind)))
```
`(comp #(when % (% text message)) kind->props-fn)` this part is a bit hard to read, can we extract it and give some name?

janherich replied:
I agree, will extract it to the separate function.

janherich commented:
@rasom - feedback incorporated into the PR

rasom commented on src/status_im/constants.cljs:
```diff
@@ -195,6 +194,9 @@
 (def regx-italic #"~[^~]+~")
 
+(def ^:const lines-collapse-threshold 20)
+(def ^:const chars-collapes-threshold 600)
```
typo `chars-collapes-threshold` should be `chars-collapse-threshold`

Serhy commented:
@janherich there is an app crash happens while message formatting for one of the following URLs (haven't found yet which one exactly but here is a text snippet to reproduce)
1) Open Status, create an account and join any chat (e.g. unique public `pull6372`) 
2) Send (or receive) the next message: 
```http://✪df.ws/123
  http://userid:password@example.com:8080
   http://userid:password@example.com:8080/
   http://userid@example.com
   http://userid@example.com/
   http://userid@example.com:8080
   http://userid@example.com:8080/
   http://userid:password@example.com
   http://userid:password@example.com/
   http://142.42.1.1/
   http://142.42.1.1:8080/
   http://:arrow_right:.ws/䨹
   http://⌘.ws
   http://⌘.ws/
   http://foo.com/blah_(wikipedia)#cite-1
   http://foo.com/blah_(wikipedia)_blah#cite-1
   http://foo.com/unicode_(✪)_in_parens
   http://foo.com/(something)?after=parens
   http://:relaxed:.damowmow.com/
   http://code.google.com/events/#&product=browser
   http://j.mp
   ftp://foo.bar/baz
   http://foo.bar/?q=Test%20URL-encoded%20stuff
   http://مثال.إختبار
   http://例子.测试

```

Actual result: No protocol method iCollection-conj defined for type function...

<img width="368" alt="screen shot 2018-10-18 at 18 28 12" src="https://user-images.githubusercontent.com/8749671/47165863-9c925b00-d303-11e8-98bd-d401d876120b.png">

I didn't see yet extra boost in app responsivness when receiving 500 text messages at once (joining public chat). On Galaxy Note 4 it's a bit better than what we have in Release 0.9.29  (Will provide numbers later in scope of issue #4193) 
But question: do we expect lags be decreased in a case when contact receives messages from offline inbox OR when navigating to a chat screen with the messages already fetched?  

Continue testing...

janherich commented:
@Serhy Thanks for the valuable input, I'm investigating/fixing already. Regarding the perf boost, it should be only related to rendering, eq when you have chat open and a lot of messages are received, or you open chat with lot of messages and scroll repeatedly to the top to have them all loaded in memory (and rendered).
No perf boost is expected when just staying at the home screen (messages are not rendered).

janherich commented:
@Serhy The issue has been already fixed.

janherich commented:
@Serhy anything new ? I added some minor changes, like fixing the collapse/expand styling for long messages (is broken in the current develop) + rendering of message styling (bold/italic) in home view as well - I intentionally didn't add actions (link/tags) to the home view, as it could interact with navigation to chat on small screens.

Serhy commented:
Sorry, @janherich I switched on 0.9.30 release related tasks, were unable to pick up latest fix here. 
Will be able to get back to it once we are done with the release.

Serhy commented:
@janherich latest builds looks good to me: no links/tags on home screen = good! **Bold**/_Italic_ fonts are reflected on home screen (fixing https://github.com/status-im/status-react/issues/3199)

**(Issue 1)** However, `/send` and `/receive` messages are not visible on Home screen if they are the latest messages sent messages in chat 

**Issue 2** After upgrade from release app version to current PR build, - all the messages in history loose the formatting (links / tags / bold /italic) - all displayed as plain text and all large messages expanded by default (with no option to collapse). Further messaging looks good and formatting takes place.
**To reproduce:**
1) Install 0.9.29 (or 0.9.30) Release app version
2) Create new account and join next public chats: `#status`, `#biglink`, `bolditalic`
3) Upgrade to https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181024-080238-9542c9-pr.apk (for Android) 
4) Log in to account and check whether formatting is kept in the chats and bigmessages (`biglink` chat) collapsed

I still need to measure to have exact figures about performance it terms of message rendering / scrolling (it's really milliseconds difference between nightly and current PR build), scrolling chat with a lot of messages and expanding huge message (e.g. the one in `#hobbit` public chat) behaves the same in timing.

<blockquote></blockquote>

janherich commented:
@Serhy Both issues are fixed in the new build

churik commented:
@cammellos one more thing - after 8 installations and 14 contacts I lost ability to sync devices (can repro from all 3 devices). Waiting/relogin doesn't help.
Video from desktop + android: http://take.ms/IE1pT
Logs:
[Status.log](https://github.com/status-im/status-react/files/2514789/Status.log)
[geth.log](https://github.com/status-im/status-react/files/2514791/geth.log)

cammellos commented:
@churik thanks, wrong github PR :) but I'll take a look at it

Serhy commented:
@janherich functional bugs are all fixed. 
I've tried to measure performance difference between latest builds in this PR (https://github.com/status-im/status-react/pull/6372#issuecomment-432809939) and latest nightly on Galaxy Note 4 (we take it as weak device, and it's on Android 6.0.1). 

I have to say we regress in particular cases for performance.

There is almost no difference between nightly and PR build when I: 
* scroll the chat with 100 simple text messages (1 message ~20 chars)
* expand/collapse big messages having text

However, when text has formatted parts it takes more time for messages to appear, expand/collapse. Please see below steps and values (note: below values are average of 3 runs):

1) Open Status (Device1=GalaxyNote4 in my case) and create new account
2) Join any unique public chat (so that we work with the single message later)
3) Navigate back to Home screen.
4) From another Device2 (device/build doesn't matter) join the same unique public chat as in 2nd item
5) Send TWO messages 3000 chars each, message content is `links` (see attachment for what I used to send)
6) After GalaxyNote4 received the message (visible on Home screen) and `Fetching messages...` is gone -> tap on public chat and pay attention for when message appears. Then once message is visible expand / collapse the message

| Action | Develop build  | [PR build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181024-194657-4b26c3-pr.apk) |
| --- | --- | --- |
| When big message with links appears |  3 sec |  6 sec  |
| Expand |  <1 sec |  3 sec  |
| Collapse |  <1 sec |  3 sec  |

7) Repeate 1-4 items above
8) Send TWO messages 3000 chars each, message content is `#tags` (see attachment for what I used to send)

9) After GalaxyNote4 received the message (visible on Home screen) and `Fetching messages...` is gone navigate to public chat and pay attention for when message appears. Expand/Collapse.

| Action | Develop build  | [PR build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181024-194657-4b26c3-pr.apk) |
| --- | --- | --- |
| When big message with TAGS appears |  <1 sec |  20 sec  |
| Expand |  <1 sec |  8.5 sec  |
| Collapse |  <1 sec |  10.3 sec  |

NOTE: difference for test with the TAGS is not valid actually because in nighlty build we don't have tags yet (they are plain text). But I just wanted to emphasise that tags VS link parsing is *heavier* for app

10) Repeate 1-4 items above
11) Send ONE message with 3000 chars, message content is `~italic text~` (see attachment for what I used to send)

| Action | Develop build  | [PR build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181024-194657-4b26c3-pr.apk) |
| --- | --- | --- |
| When big message with italic formatting appears |  6.5 sec |  8 sec  |
| Expand |  <4 sec |  6 sec  |
| Collapse |  <4 sec |  5.5 sec  |

NOTE: for italic text scanario I barely tried to get the figures. It was very often when app crashed:

<img width="368" alt="screen shot 2018-10-25 at 17 47 13" src="https://user-images.githubusercontent.com/8749671/47509099-0a97ce80-d87e-11e8-8fab-162c78c8cf4c.png">
(in process now to get the logs)

Additional side effect of performance regress is navigation between main tabs: apparently because we apply formatting on Home screen as well, it also takes more time and bugger lags when I navigate to Howe screen with the very last messaeg in any chat - is a big message (for above scenarios, for example with _~italic text~_ - 8 seconds to load Home screen with a single public chat VS 1 second in Nightly build (8 times more!)

These are figures I could get. Again, for text messages it's almost no difference (entire 'Hobbit' book in a single message expands/collapses/appears pretty fast (<2s) with the similar values in nightly/currentBuild), but parsing formatted text brings us regression and app more laggy than what is in develop.

Single message content used in above scenarios:
[italic_3000_chars.txt](https://github.com/status-im/status-react/files/2515552/italic_3000_chars.txt)
[links_3000_chars.txt](https://github.com/status-im/status-react/files/2515553/links_3000_chars.txt)
[tags_3000_chars.txt](https://github.com/status-im/status-react/files/2515554/tags_3000_chars.txt)

janherich commented:
Thanks for the valuable insights @Serhy, especially the perf testing - in light of that, I will abandon hierarchic parsing (so for example you can have formatted links/tags, or nested formating, like ~*bold-and-italic*~) because it's actually too much for slower devices, and will go with simple flat formatting, where any piece of text is allowed to only have one "modifier" (action of formatting).
This will make the code actually faster then current develop.

janherich commented:
@Serhy I fixed the perf regressions and reverted to one-level formatting/actions, so any part of text can be formatted (bold/italic) or action (tag/link) but not both/nested just as it's now in develop. 
Let's see how it now compares performance wise with the develop :)

Serhy commented:
@janherich Builds from https://github.com/status-im/status-react/pull/6372#issuecomment-434241047 on iPhoneX and Samsung Galaxy J7 with the same load as in the https://github.com/status-im/status-react/pull/6372#issuecomment-433089394

**Results**:

| Action | Develop build  | [PR build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181030-094342-9f4e16-pr.apk) |
| --- | --- | --- |
| When 2 big messages (3k chars each) with TAGS appears | <1 sec |  8 sec  |
| Expand |  <1 sec |  1.8 sec  |
| Collapse |  <1 sec |  3.6 sec  |

Note: same to previous comment, - can't compare to nightly because we don't have tags there (they are plain text) but, - it's three times better than what was in previous build.

| Action | Develop build  | [PR build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181030-094342-9f4e16-pr.apk) |
| --- | --- | --- |
| When 2 big messages (3k chars each) with LINKS appears | 2.4 sec |  3.4 sec  |
| Expand |  <1 sec |  1 sec  |
| Collapse |  <1 sec |  1.5 sec  |

| Action | Develop build  | [PR build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181030-094342-9f4e16-pr.apk) |
| --- | --- | --- |
| When message with 3k chars with ITALIC (or BOLD) font appears | 2.4 sec |  NaN sec  |
| Expand |  <1 sec |  1.5 sec  |
| Collapse |  <1 sec |  2 sec  |

In average, it's three times better than what we had with previous changes for tags/links (comment https://github.com/status-im/status-react/pull/6372#issuecomment-433089394). 
It seem nothing changed for how long it takes to navigate to the chat with italic / bold font text: so 3k chars (350 separated words with bold / italic font) takes way too long: from 4 tests I had 3 crashes when navigating to chat screen, and 1 time I could get to chat screen only after submitted "Wait more" on app 'App is not responds' pop-up. 

What we have in develop still wins in performance for text with formatting.

janherich commented:
@Serhy In the newest build, I tried to improve the performance by moving as much processing as possible into the background ("fetching messages") steps and not rendering.
It works fine on the Huawei P9 (Android 8.0.0) but I don't have slower device available.

Serhy commented:
@janherich latest builds (posted in https://github.com/status-im/status-react/pull/6372#issuecomment-435421600) retested on Galaxy Note 4 for the receiving, expanding, collapsing of big messages with links/tags/bold/italic items in text:


 | Action | Develop build  | [PR build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181102-135724-fcaf29-pr.apk) |
| --- | --- | --- |
| When 2 big messages (3k chars each) with LINKS appears | ~2.5 sec |  4 sec  |
| Expand |  <1 sec |  1.5 sec  |
| Collapse |  <1 sec |  1.5 sec  |

 | Action | Develop build  | [PR build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181102-135724-fcaf29-pr.apk) |
| --- | --- | --- |
| When 2 big messages (3k chars each) with TAGS appears | <1 sec (no parsing takes place  here so it's OK) |  12 sec  |
| Expand |  <1 sec |  5 sec  |
| Collapse |  <1 sec |  5 sec  |

 | Action | Develop build  | [PR build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181102-135724-fcaf29-pr.apk) |
| --- | --- | --- |
| When 1 big message (3k chars, 350 words) with ITALIC font appears | 4 sec |  4 sec  |
| Expand |  2.8 sec |  2.8 sec  |
| Collapse |  2.8 sec |  2.8 sec  |

 | Action | Develop build  | [PR build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181102-135724-fcaf29-pr.apk) |
| --- | --- | --- |
| When 1 big message (3k chars, 350 words) with BOLD font appears | 4 sec |  4 sec  |
| Expand |  2.7 sec |  2.5 sec  |
| Collapse |  2.8 sec |  2.5 sec  |

Nothing changed when scrolling chat with 100 messages with mixed content (text + links + markup).

**Note**: there is no performance impact when navigating to Home screen on Mobile device (e.g. Wallet-Home-Profile-Home) that was present before, which is great!
I see no performance improvement on mobile nor degradation noticeable to user. 

However things got a bit worse on Desktop client (checked on MacOS):

The text markup (italic / bold) now present on Desktop but only for **bold** markup (fixes partially issue https://github.com/status-im/status-react/issues/5275). While no italic font displayed on Desktop for text wrapped up in tildes, - it's visible that text parsing takes place (visible by performance impact + no tildes displayed for text -> similar to #6361 on mobile)

Comparing to latest develop build for desktop MacOS client:

 | Action | Develop build  | [PR build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181102-135724-fcaf29-pr.dmg) |
| --- | --- | --- |
| When 1 big message (3k chars, 350 words) with BOLD / ITALIC font appears | <1 sec (because no parsing) |  30 sec  |

| Action | Develop build  | [PR build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181102-135724-fcaf29-pr.dmg) |
| --- | --- | --- |
| When 1 big message (3k chars, 350 words) with LINKS font appears | 5 |  10 sec  |

| Action | Develop build  | [PR build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181102-135724-fcaf29-pr.dmg) |
| --- | --- | --- |
| When 1 big message (3k chars, 350 words) with TAGS font appears | 10 | 21 sec  |

We regress twice for above cases - it would be nice to apply the same (or similar) logic for message content parsing on Desktop like you've done on mobile.

Also, two functional issues on desktop appeared:
1) Link became clickable in chat on Desktop. But when I click any link - the `Can not read property 'showPicker' of unidentified` appears:

<img width="626" alt="screen shot 2018-11-05 at 13 23 11" src="https://user-images.githubusercontent.com/8749671/48000231-89f29100-e10d-11e8-8398-43c1287a507a.png">

**NOTE**: I see it was fixed https://github.com/status-im/status-react/issues/6565, so might rebase develop could would fix it for us! 

2) It was a good idea to not make links / tags in chat list actionable on mobile (user need to read the message content first in order to decide whether to click or not click on item). Lets make the same on Desktop: at the moment when the beginning of the latest message in chat consist a link or a tag - it could be clicked and `Can not read property 'showPicker' of unidentified` error appears for link (above item) or public chat created (for Tags).

<img width="483" alt="screen shot 2018-11-05 at 15 34 21" src="https://user-images.githubusercontent.com/8749671/48001259-b360ec00-e110-11e8-97e3-98acb86fe07e.png">

janherich commented:
@Serhy all issues are fixed in the newest build (coming soon)

Serhy commented:
All functional bugs are fixed, @janherich, but are we able to optimise parsing of content messages on desktop client in the scope of this PR?
It still behaves same to what was before (in https://github.com/status-im/status-react/pull/6372#issuecomment-435874637 for desktop) - so big message having only links takes twice more time in order for a particular message to appear (and in general freeze the app). However, there is no difference in scrolling the chat with a lot of messages with mixed content (text + emoji + links).

janherich commented:
@Serhy The desktop regression should be fixed in the latest build (will be ready soon)

Serhy commented:
All looks good with the latest changes, @janherich !
Failed autotests rechecked (not this PR issues as they failed in latest nightly too)

It’s worth to note that now message rendering time even more improved on mobile! For the Samsung Galaxy Note 4:

 | Action | Develop build  | [PR build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181106-135540-079800-pr.apk) |
| --- | --- | --- |
| When 2 big messages (3k chars each) with LINKS appears | ~2.5 sec |  <2.5 sec  |
| Expand |  <1 sec |  <1 sec  |
| Collapse |  <1 sec |  <1 sec  |

 | Action | Develop build  | [PR build](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-181106-135540-079800-pr.apk) |
| --- | --- | --- |
| When 1 big messages (350 words each wrapped in asterisk) with BOLD font appears | ~2.5 sec |  <2 sec  |
| Expand |  <1 sec |  <1 sec  |
| Collapse |  <1 sec |  <1 sec  |

Additionally:
* I don’t see https://github.com/status-im/status-react/issues/5960 issue on Desktop happening anymore (only MacOS build checked so far)
* navigating to public chat links on mobile redirects to respective public chat (joins if it did not exist before)
* half fixes https://github.com/status-im/status-react/issues/5275 introducing bold font on desktop
